[NWC] Add an encryption tag to negotiate upgrading to NIP44. (#1780)
nostr-protocol/nips: 47.md

 
 4. Once the payment is complete the **wallet service** will send an encrypted `response` (kind 23195) to the **user** over the relay(s) in the URI.
 
 5. The **wallet service** may send encrypted notifications (kind 23196) of wallet events (such as a received payment) to the **client**.
 5. The **wallet service** may send encrypted notifications (kind 23197 or 23196) of wallet events (such as a received payment) to the **client**.

## Events

There are four event kinds:

- `NIP-47 info event`: 13194
- `NIP-47 request`: 23194
- `NIP-47 response`: 23195

> - `NIP-47 notification event`: 23196

Reviewer: What is the downside of just keeping the 23196 kind notification type and also add an "encryption" tag to be consistent with the way it's done for request and response events as well? What is the reason a new kind 23197 is needed?

Reviewer 2: same question here

Reviewer 3: If we do this then all the clients currently using NIP-04 without checking for the encryption tag will try to decrypt NIP-44 encrypted notifications, this helps with backwards compatibility

Author: Yep @im-adithya beat me to it :-). For backwards-compatibility, it's easier for folks who don't know about this upgrade to just continue to do the same thing they've been doing. In an ideal world, maybe we could just send 2 23196 events with each of the encryption types, but I don't want to cause too many weird errors for legacy apps trying to decrypt nip44 as nip04.

        // ...
    ],
    "content": "pay_invoice get_balance make_invoice lookup_invoice list_transactions get_info notifications",
    // ...
}
```

### Request and Response Events

Both the request and response events SHOULD contain one `p` tag, containing the public key of the **wallet service** if this is a request, and the public key of the **client** if this is a response. The response event SHOULD contain an `e` tag with the id of the request event it is responding to.
Optionally, a request can have an `expiration` tag that has a unix timestamp in seconds. If the request is received after this timestamp, it should be ignored.

The content of requests and responses is encrypted with [NIP44](44.md), and is a JSON-RPCish object with a semi-fixed structure.

**Important note for backwards-compatibility:** The initial version of the protocol used [NIP04](04.md). If a **wallet service** or client app does not include the `encryption` tag in the
`info` or request events, it should be assumed that the connection is using NIP04 for encryption. See the [Encryption](#encryption) section for more information.

Example request:

```js
{
    "kind" 23194,
    "tags": [

>         ["encryption", "nip44"],

Reviewer 4: nip44 is a versioned protocol which might cause problems in the future. Should we use `nip44_v2` instead of `nip44`?

Author: Good catch!

        ["p", "03..." ] // public key of the wallet service.
        // ...
    ],

>     "content": nip44_encrypt({

Reviewer 2: Should this also indicate here that this depends on the `encryption` tag?

It's mentioned in the section about encryption though so maybe that's enough?

> The **wallet service** and **client** should negotiate the encryption method to use based on the `encryption` tags in the `info` event.

Author: Added a comment

        "method": "pay_invoice", // method, string
        "params": { // params, object
            "invoice": "lnbc50n1..." // command-related data
        }
    }),
}
```

Example response:

```js
{
    "kind" 23195,
    "tags": [

>         ["encryption", "nip44_v2"],

Reviewer 3: Can we remove/make this optional as we usually expect this to be the same as that of `request_event`?

Author: I was leaning towards being explicit here, but you're right that it's not really necessary. Will remove.

### Notification Events

> The notification event is a kind 23197 event (23196 if using nip04) SHOULD contain one `p` tag, the public key of the **user**.

Reviewer 4: ```suggestion
The notification event is a kind 23197 event (23196 if using nip04) SHOULD contain one `p` tag, the public key of the **client**.
```

Reviewer 4: Also I'm wondering if we even mention 23196 here since we have the note on backward compatibility below?

Author: Done and removed that reference too.

}
```

> _Note on backwards-compatibility:_ If a **wallet service** supports both nip44 and nip04 for legacy client apps, it should publish both notification events for each notification - kind 23196 encrypted with NIP-04, and kind 23197 encrypted with NIP-44. It is up to the **client** to decide which event to listen to based on its supported encryption and declared supported encryption schemes of the **wallet service** in the `info` event.

Reviewer 4: Is this note needed since there's a backward compatibility section further down?

Author: I thought it was helpful in-context, but you're right it's not strictly necessary.

- `RESTRICTED`: This public key is not allowed to do this operation.
- `UNAUTHORIZED`: This public key has no wallet connected.
- `INTERNAL`: An internal error.

> - `UNSUPPORTED_ENCRYPTION`: The encryption type of the request is not supported by the wallet service.

Reviewer 3: Will this be used? If yes, how will this error be encrypted? :smile:

I think it's better to not respond to such requests to reduce complexity...

Author: Good catch on the encryption point! I'd kind of lost the fact that errors also need to be encrypted. Thinking through this a bit, I'm not sure I love not responding to these events because that's not very debuggable on the client app side and will leave client apps hanging until they time out, so it's not a great ux either. We can't encrypt the error using the client's encryption type because by definition in this case the wallet service doesn't support it.

I think the best behavior here is actually to just encrypt the error content using the wallet service's default encryption scheme. Of course the client won't be able to decrypt it, but at least they'll get a decryption error while debugging, which will tip them in the right direction. There may even be some cases where it was actually just the client's encryption tag which was wrong, and so they will see the error message. For example:

1) Client forgets to populate the encryption tag in a request or sends something like `nip44` or `nip44v2` instead of `nip44_v2`
2) Wallet service sees this and doesn't understand, so it responds with a nip44 encrypted UNSUPPORTED_ENCRYPTION error
3) Client app can actually still decrypt this and see the right error type to nudge them to re-check their encryption tags.

Does that make sense to you? If so, I can add some clarification to this PR.

Reviewer 4: On the other hand, building a lot of extra logic to try to catch all the cases where a developer doesn't implement the spec properly is a slippery slope and makes more complicated code to account for all these cases as Adithya mentioned. But it's a good point that developers who cannot see the wallet service logs will find it harder to debug. In Alby Hub the developer can see the logs, but this won't be the case for services where the developer is not the owner of the backend.

Author: After thinking about this more, I still feel pretty strongly that the debugging experience would be pretty painful if servers just ignore these messages entirely. As a client app developer, it's pretty likely that I didn't write the software for my wallet and that I'm not running it locally. It feels like a pretty easy foot-gun to write `nip44` instead of `nip44_v2` and spend the next several hours confused why your app doesn't work. IMO being defensive here is just good devX.

I do understand your point about not trying to catch every case where a developer doesn't implement the spec correctly, but in this case, even doing nothing is a choice in special handling. Even then you have already looked at the message, seen an invalid encryption scheme, and then decided to do nothing instead of returning an error. I don't think either case is really much extra handling tbh.

2. **wallet service** verifies that the author's key is authorized to perform the payment, decrypts the payload and sends the payment.
3. **wallet service** responds to the event by sending an event with kind `23195` and content being a response either containing an error message or a preimage.

## Encryption

The initial version of NWC used [NIP-04](04.md) for encryption which has been deprecated and replaced by [NIP-44](44.md). NIP-44 should always be preferred for encryption, but there may be legacy cases

> where wither the **wallet service** or **client** has not yet migrated to NIP-44. The **wallet service** and **client** should negotiate the encryption method to use based on the `encryption` tags in the `info` event.

Reviewer: Just a typo I think, "wither" should be "whether".

Author: Thanks :-)

> where wither the **wallet service** or **client** has not yet migrated to NIP-44. The **wallet service** and **client** should negotiate the encryption method to use based on the `encryption` tags in the `info` event.

Reviewer 2: ```suggestion
where wether the **wallet service** or **client** has not yet migrated to NIP-44. The **wallet service** and **client** should negotiate the encryption method to use based on the `encryption` tag in the `info` event.
```

### Request events

When a **client application** sends a request event, it should include a `encryption` tag with the encryption scheme it is using. The scheme MUST be supported by the **wallet service** as indicated by the info event.
For example, if the client application supports nip44, the request event might look like:

```jsonc
{
    "kind": 23194,
    "tags": [
        ["encryption", "nip44_v2"],
        // ...
    ],
    // ...
}
```

If the **wallet service** does not support the specified encryption scheme, it will return an `UNSUPPORTED_ENCRYPTION` error. Absence of the `encryption` tag indicates use of nip04 for encryption.

### Notification events

As described above in the [Notifications](#notifications) section, if a **wallet service** supports both nip04 and nip44, it should publish two notification events for each notification - kind 23196 encrypted with NIP-04, and kind 23197 encrypted with NIP-44. If the **wallet service** only supports nip44, it should only publish kind 23197 events.

> The **client** should check the `encryption` tag in the `info` event to determine which encryption schemes the **wallet service** supports, and listen to the appropriate notification event. Clients that support both nip04 and nip44 can choose to only listen to 23197 events for simplicity.

Reviewer 4: ```suggestion
The **client** should check the `encryption` tag in the `info` event to determine which encryption schemes the **wallet service** supports, and listen to the appropriate notification event.
```
